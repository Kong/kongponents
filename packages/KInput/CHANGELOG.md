# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.2.2](https://github.com/Kong/kongponents/compare/@kongponents/kinput@0.2.0-0...@kongponents/kinput@0.2.2) (2020-01-16)

**Note:** Version bump only for package @kongponents/kinput





# [0.2.0-0](https://github.com/Kong/kongponents/compare/@kongponents/kinput@0.1.3-5...@kongponents/kinput@0.2.0-0) (2019-11-14)


### Features

* **forms:** Add KRadio & KCheckbox, add form css ([65158f2](https://github.com/Kong/kongponents/commit/65158f26ca6999a6fe47e0c58fe5a64aed8737d6))





## [0.1.3-5](https://github.com/Kong/kongponents/compare/@kongponents/kinput@0.1.3-4...@kongponents/kinput@0.1.3-5) (2019-11-12)

**Note:** Version bump only for package @kongponents/kinput





## [0.1.3-4](https://github.com/Kong/kongponents/compare/@kongponents/kinput@0.1.3-3...@kongponents/kinput@0.1.3-4) (2019-11-11)

**Note:** Version bump only for package @kongponents/kinput





## [0.1.3-3](https://github.com/Kong/kongponents/compare/@kongponents/kinput@0.1.3-2...@kongponents/kinput@0.1.3-3) (2019-11-08)


### Bug Fixes

* **dist:** dist to all packages ([#152](https://github.com/Kong/kongponents/issues/152)) ([217c46b](https://github.com/Kong/kongponents/commit/217c46b02afe960f8fa597ba77410dcf034b1716))





## [0.1.3-2](https://github.com/Kong/kongponents/compare/@kongponents/kinput@0.1.3-1...@kongponents/kinput@0.1.3-2) (2019-11-08)

**Note:** Version bump only for package @kongponents/kinput
