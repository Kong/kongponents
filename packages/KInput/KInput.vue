<template>
  <input
    class="form-control k-input"
    v-on="listeners">
</template>

<script>
export default {
  name: 'KInput',
  computed: {
    listeners () {
      return {
        ...this.$listeners
      }
    }
  }
}
</script>

<style lang="scss">
input.k-input + .hint {
  margin: var(--spacing-xs) 0 0;
  font-size: var(--type-sm);
  color: var(--tblack-45);
}
</style>

<style scoped lang="scss">
input {
  display: block;
  padding: var(--spacing-xs) var(--spacing-sm);
  color: var(--KInputColor, var(--tblack-70));
  font-size: var(--type-md);
  font-weight: normal;
  line-height: 1.25;
  border: 1px solid var(--KInputBorder, var(--grey-88));
  background-color: var(--KInputBackground, var(--twhite-1));
  border-radius: 3px;
}

input:focus {
  outline: none !important;
  border: 1px solid var(--KInputFocus, var(--blue-light));
}

input:disabled {
  cursor: not-allowed;
  font-style: italic;
  background-color: var(--KInputDisabledBackground, var(--grey-98));
  opacity: 0.5;
}

input:invalid,
input.input-error {
  outline: none !important;
  border: 1px solid var(--KInputError, var(--red-light));
}

input[type="search"] {
  padding-left: 36px;
  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='16' height='16'%3E%3Cpath fill='%23000' fill-opacity='.45' fill-rule='evenodd' d='M6 12c-3.3137085 0-6-2.6862915-6-6s2.6862915-6 6-6 6 2.6862915 6 6c0 1.29583043-.410791 2.49571549-1.1092521 3.47653436l1.2305724 1.23057244 2.8232632 2.8338633c.3897175.3911808.3947266 1.0192147.005164 1.4087774-.3868655.3868655-1.014825.3873148-1.4087774-.005164l-2.8338633-2.8232632-1.23057244-1.2305724C8.49571549 11.589209 7.29583043 12 6 12zm4-6c0-2.209139-1.790861-4-4-4S2 3.790861 2 6s1.790861 4 4 4 4-1.790861 4-4z'/%3E%3C/svg%3E");
  background-repeat: no-repeat;
  background-position: 12px 50%;
}

/* Browser Overrides */
input::placeholder {
  color: var(--tblack-45);
  font-weight: normal;
  opacity: 1;
}

input:invalid,
input:-moz-submit-invalid,
input:-moz-ui-invalid {
  box-shadow: none;
}

input::-ms-clear {
  display: none;
}
</style>
