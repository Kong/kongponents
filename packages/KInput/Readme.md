```js
<KInput type="text" placeholder="placeholder" />
```
