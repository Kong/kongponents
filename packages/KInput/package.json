{
  "name": "@kongponents/kinput",
  "version": "0.1.3",
  "description": "Input component",
  "main": "KInput.vue",
  "scripts": {
    "test": "echo \"Error: no test specified\""
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme"
}
