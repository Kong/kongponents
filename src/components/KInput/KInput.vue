<template>
  <div
    class="k-input"
    :class="[attrs.class, { 'input-error': charLimitExceeded || error || hasError }]"
  >
    <KLabel
      v-if="label"
      :for="inputId"
      v-bind="labelAttributes"
      :required="isRequired"
    >
      {{ strippedLabel }}

      <template
        v-if="hasLabelTooltip"
        #tooltip
      >
        <slot name="label-tooltip" />
      </template>
    </KLabel>

    <div
      class="input-element-wrapper"
      :class="{ 'has-before-content': slots.before, 'has-after-content': slots.after || (type === 'password' && showPasswordMaskToggle) }"
    >
      <div
        v-if="slots.before"
        ref="beforeSlotElement"
        class="before-content-wrapper"
      >
        <slot name="before" />
      </div>

      <input
        :id="inputId"
        ref="inputElement"
        :aria-describedby="helpText || slots.help ? helpTextId : undefined"
        :aria-invalid="error || hasError || charLimitExceeded ? 'true' : undefined"
        class="input"
        :type="inputType"
        v-bind="modifiedAttrs"
        :value="getValue()"
        @input="handleInput"
      >

      <div
        v-if="slots.after || (type === 'password' && showPasswordMaskToggle)"
        ref="afterSlotElement"
        class="after-content-wrapper"
      >
        <button
          v-if="type === 'password' && showPasswordMaskToggle"
          :aria-label="`${maskValue ? 'Hide' : 'Show'} value`"
          class="mask-value-toggle-button"
          type="button"
          @click.stop="maskValue = !maskValue"
          @mousedown.prevent
          @mouseup.prevent
        >
          <VisibilityOffIcon
            v-if="maskValue"
            decorative
          />
          <VisibilityIcon
            v-else
            decorative
          />
        </button>

        <slot
          v-else
          name="after"
        />
      </div>
    </div>

    <Transition
      mode="out-in"
      name="kongponents-fade-transition"
    >
      <p
        v-if="helpText || slots.help"
        :id="helpTextId"
        :key="String(helpTextKey)"
        class="help-text"
      >
        <slot
          v-if="showHelp"
          name="help"
        >
          {{ helpText }}
        </slot>
        <template v-else>
          {{ helpText }}
        </template>
      </p>
    </Transition>
  </div>
</template>

<script setup lang="ts">
import { computed, ref, watch, useAttrs, onMounted, nextTick, useId, useTemplateRef } from 'vue'
import type { InputProps, InputEmits, InputSlots } from '@/types'
import useUtilities from '@/composables/useUtilities'
import KLabel from '@/components/KLabel/KLabel.vue'
import { KUI_ICON_SIZE_40 } from '@kong/design-tokens'
import { VisibilityIcon, VisibilityOffIcon } from '@kong/icons'

defineOptions({
  inheritAttrs: false,
})

const {
  modelValue = '',
  label = '',
  labelAttributes = {},
  help = '',
  error,
  errorMessage = '',
  characterLimit = null,
  hasError,
  type = 'text',
  showPasswordMaskToggle,
} = defineProps<InputProps>()

watch(() => hasError, (val) => {
  if (val) {
    console.warn('KInput: `hasError` prop is deprecated. Please use `error` prop instead. See the migration guide for more details: https://kongponents.konghq.com/guide/migrating-to-version-9.html#kinput')
  }
}, { immediate: true })

watch(() => labelAttributes.help, (help) => {
  if (help) {
    console.warn('KInput: `help` property of `labelAttributes` prop is deprecated. Please use `info` prop instead. See the migration guide for more details: https://kongponents.konghq.com/guide/migrating-to-version-9.html#klabel')
  }
})

const emit = defineEmits<InputEmits>()
const slots = defineSlots<InputSlots>()

const inputElementRef = useTemplateRef('inputElement')

const currValue = ref<string>('') // We need this so that we don't lose the updated value on hover/blur event with label
const modelValueChanged = ref<boolean>(false) // Determine if the original value was modified by the user
const helpTextKey = ref<number>(0)

const { stripRequiredLabel } = useUtilities()
const attrs = useAttrs()

const isRequired = computed((): boolean => attrs?.required !== undefined && String(attrs?.required) !== 'false')
const defaultId = useId()
const inputId = computed((): string => attrs.id ? String(attrs.id) : defaultId)
const helpTextId = useId()
const strippedLabel = computed((): string => stripRequiredLabel(label, isRequired.value))
const hasLabelTooltip = computed((): boolean => !!(labelAttributes?.info || slots['label-tooltip']))

// we need this so we can create a watcher for programmatic changes to the modelValue
const value = computed({
  get(): string | number {
    return modelValue
  },
  set(newValue: string | number): void {
    // @ts-ignore: allow typing as Event
    handleInput({ target: { value: newValue } } as Event)
  },
})

const modifiedAttrs = computed((): Record<string, any> => {
  const $attrs = { ...attrs }

  // delete classes because we bind them to the parent
  delete $attrs.class
  // use @input in template for v-model support
  delete $attrs.input
  delete $attrs.onInput

  return $attrs
})

const charLimitExceeded = computed((): boolean => {
  const currValLength = currValue.value?.toString().length || 0
  const modelValLength = modelValue?.toString().length || 0

  // default to length of currVal
  let length = currValLength

  // if there is a model value and it hasn't been modified yet, use that instead
  if (!modelValueChanged.value && modelValLength) {
    length = modelValLength
  }

  return !!characterLimit && length > characterLimit
})

const charLimitExceededErrorMessage = computed((): string => {
  if (!charLimitExceeded.value) {
    return ''
  }

  return modelValueChanged.value
    ? `${currValue.value?.toString().length} / ${characterLimit}`
    : `${modelValue?.toString().length} / ${characterLimit}`
})

// Whether `help` slot or `help` prop should be shown
const showHelp = computed((): boolean => !charLimitExceeded.value && !error && !hasError && !!(help || slots.help))

const helpText = computed((): string => {
  // if character limit exceeded, return that error message
  if (charLimitExceeded.value) {
    return charLimitExceededErrorMessage.value
  }

  // if error prop is true and there is an error message, return that
  if ((error || hasError) && errorMessage) {
    return errorMessage
  }

  // otherwise return the help text
  // if error prop is true it danger styles will be applied
  return help
})

watch(charLimitExceeded, (newVal, oldVal) => {
  if (newVal !== oldVal) {
    emit('char-limit-exceeded', {
      value: currValue.value,
      length: currValue.value.length,
      characterLimit: characterLimit!,
      limitExceeded: newVal,
    })

    // bump the key to trigger the transition
    helpTextKey.value += 1
  }
})

watch(value, (newVal, oldVal) => {
  if (newVal !== oldVal) {
    // @ts-ignore: allow typing as Event
    handleInput({ target: { value: newVal } } as Event)
  }
})

const handleInput = (event: Event): void => {
  // avoid pass by ref
  const value = JSON.parse(JSON.stringify((event?.target as HTMLInputElement)?.value))

  updateInputValue(value)
}

const updateInputValue = (value: string): void => {
  currValue.value = value
  modelValueChanged.value = true

  emit('input', value)
  emit('update:modelValue', value)
}

const getValue = (): string | number => {
  // Use the modelValue only if it was initialized to something and the value hasn't been changed
  return currValue.value || modelValueChanged.value ? currValue.value : modelValue
}

watch(() => error, (newVal, oldVal) => {
  if (newVal !== oldVal) {
    // bump the key to trigger the transition
    helpTextKey.value += 1
  }
})

const beforeSlotElement = ref<HTMLElement | null>(null)
const afterSlotElement = ref<HTMLElement | null>(null)
const beforeSlotElementWidth = ref<string>(KUI_ICON_SIZE_40) // default to slot icon size
const afterSlotElementWidth = ref<string>(KUI_ICON_SIZE_40) // default to slot icon size

const maskValue = ref<boolean>(false)
const inputType = computed((): string => {
  return type === 'password' && maskValue.value ? 'text' : type
})

onMounted(async () => {
  await nextTick() // wait for the slots content to render

  if (beforeSlotElement.value?.offsetWidth) {
    beforeSlotElementWidth.value = beforeSlotElement.value.offsetWidth + 'px'
  }

  if (afterSlotElement.value?.offsetWidth) {
    afterSlotElementWidth.value = afterSlotElement.value.offsetWidth + 'px'
  }
})

defineExpose({
  input: inputElementRef,
})
</script>

<style lang="scss" scoped>
/* Component variables */
// Only add variables here sparingly for ease of use when the same value needs to be referenced for display logic.

$kInputPaddingX: var(--kui-space-50, $kui-space-50); // corresponds to mixin, search for variable name in mixins
$kInputIconSize: var(--kui-icon-size-40, $kui-icon-size-40); // $kSelectInputIconSize
$kInputSlotSpacing: var(--kui-space-40, $kui-space-40); // $kSelectInputSlotSpacing

/* Component styles */

.k-input {
  display: flex;
  flex-direction: column;
  width: 100%;

  // error styles
  &.input-error {
    .input, .input[type="file"] {
      @include inputError;

      &:hover {
        @include inputErrorHover;
      }

      &:focus {
        @include inputErrorFocus;
      }
    }

    .help-text {
      color: var(--kui-color-text-danger, $kui-color-text-danger);
    }
  }

  .help-text {
    @include inputHelpText;

    // fixing mixed-decls deprecation: https://sass-lang.com/d/mixed-decls
    // stylelint-disable-next-line no-duplicate-selectors
    & {
      // reset default margin from browser
      margin: 0;
      margin-top: var(--kui-space-40, $kui-space-40) !important; // need important to override some overrides of default p margin in other components
    }
  }

  // slots styles
  .input-element-wrapper {
    position: relative;

    .before-content-wrapper,
    .after-content-wrapper {
      color: var(--kui-color-text-neutral, $kui-color-text-neutral);
      display: inline-flex;
      gap: var(--kui-space-10, $kui-space-10);
      position: absolute;
      top: 50%;
      transform: translateY(-50%);

      // enforce icon size exported by @kong/icons because it's defined by the design system
      :deep(#{$kongponentsKongIconSelector}) {
        height: $kInputIconSize !important;
        width: $kInputIconSize !important;
      }

      // enhance the experience for most common cases that icon only slots should not
      // prevent the input from being focused by click on the icon
      &:has(> #{$kongponentsKongIconSelector}:not(button):not([role="button"]):only-child) {
        pointer-events: none;
      }

      :deep([role="button"]:not(.k-button)), :deep(button:not(.k-button)),
      .mask-value-toggle-button {
        @include defaultButtonReset;

        // fixing mixed-decls deprecation: https://sass-lang.com/d/mixed-decls
        // stylelint-disable-next-line no-duplicate-selectors
        & {
          color: var(--kui-color-text-neutral, $kui-color-text-neutral);
        }

        &:not([disabled]) {
          border-radius: var(--kui-border-radius-20, $kui-border-radius-20);
          cursor: pointer;
          outline: none;

          &:hover, &:focus, &:focus-visible {
            color: var(--kui-color-text, $kui-color-text) !important;
          }

          &:focus-visible {
            box-shadow: var(--kui-shadow-focus, $kui-shadow-focus);
          }
        }

        &[disabled] {
          color: var(--kui-color-text-disabled, $kui-color-text-disabled) !important;
          pointer-events: none;
        }
      }
    }

    .before-content-wrapper {
      left: 0;
      margin-left: $kInputPaddingX;
    }

    .after-content-wrapper {
      margin-right: $kInputPaddingX;
      right: 0;
    }

    &.has-before-content {
      .input {
        // if there is a before slot, add padding to the left of the input
        // standard padding + slot with + space between icon and input
        padding-left: calc($kInputPaddingX + v-bind('beforeSlotElementWidth') + $kInputSlotSpacing);
      }
    }

    &.has-after-content {
      .input {
        // if there is a after slot, add padding to the right of the input
        // standard padding + slot with + space between icon and input
        padding-right: calc($kInputPaddingX + v-bind('afterSlotElementWidth') + $kInputSlotSpacing);
      }
    }
  }
}

.input {
  @include inputDefaults;

  &:hover {
    @include inputHover;
  }

  &:focus {
    @include inputFocus;
  }

  &:disabled {
    @include inputDisabled;
  }

  &:read-only {
    @include inputReadOnly;

    // by default type="file" is read-only so we need to apply default input styles to override read-only styles
    &[type="file"] {
      @include inputDefaults;

      // fixing mixed-decls deprecation: https://sass-lang.com/d/mixed-decls
      // stylelint-disable-next-line no-duplicate-selectors
      & {
        cursor: pointer;
      }

      &:hover {
        @include inputHover;
      }

      &:focus {
        @include inputFocus;
      }

      &:disabled {
        @include inputDisabled;
      }
    }
  }

  // hide default password reveal icon in Edge
  &::-ms-reveal {
    display: none;
  }
}
</style>
