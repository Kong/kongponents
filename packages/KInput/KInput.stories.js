import { storiesOf } from '../../storybook/node_modules/@storybook/vue'
import { action } from '../../storybook/node_modules/@storybook/addon-actions'
import KInput from './KInput.vue'
import KLabel from '../KLabel/KLabel.vue'

storiesOf('KInput', module)
  .add('Default', () => ({
    components: { KInput, KLabel },
    template: `<div style="display:grid; grid-gap: 1rem">
      <div><KInput type="email" placeholder="Enter email" value="invalid" /></div>
      <div>
        <KLabel>Label</KLabel>
        <KInput type="text" placeholder="Focus on me" />
      </div>
      <div>
        <KLabel>Label</KLabel>
        <KInput type="text" placeholder="I have a label & hint" />
        <p class="hint">Hint</p>
      </div>
      <div>
        <KInput type="search" placeholder="Search..." />
      </div>
      <div>
        <KInput disabled type="text" placeholder="Disabled" />
      </div>
    </div>`
  }), { info: {} })
  .add('Actions', () => ({
    components: { KInput },
    template: `
      <KInput
        type="email"
        placeholder="Make some actions ðŸ˜‰"
        @focus="onFocus"
        @blur="onBlur"
        @input="onInput"/>
    `,
    methods: {
      onFocus (e) {
        action('onFocus')(e)
      },
      onBlur (e) {
        action('onBlur')(e)
      },
      onInput (e) {
        action('onInput')(e)
      }
    }
  }), { info: {} })
