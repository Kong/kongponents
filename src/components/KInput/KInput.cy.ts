import { mount } from 'cypress/vue'
import KInput from '@/components/KInput/KInput.vue'
import { h } from 'vue'

describe('KInput', () => {
  it('renders text when value is passed', () => {
    const text = 'Hello'

    mount(KInput, {
      props: {
        modelValue: text, // e.g. v-model
      },
    })

    cy.get('input').should('have.value', text)
  })

  it('renders `null` modelValue as empty string', () => {
    // @ts-ignore - to allow passing an invalid modelValue
    mount(KInput, {
      props: {
        modelValue: null, // e.g. v-model
      },
    })

    cy.get('input').should('not.have.value', 'null')
    cy.get('input').should('have.value', '')
  })

  it('renders `undefined` modelValue as empty string', () => {
    mount(KInput, {
      props: {
        modelValue: undefined, // e.g. v-model
      },
    })

    cy.get('input').should('not.have.value', 'undefined')
    cy.get('input').should('have.value', '')
  })

  it('renders label when value is passed', () => {
    const label = 'A label'

    mount(KInput, {
      props: {
        label,
      },
    })

    cy.get('.k-label').should('contain.text', label)
  })

  it('renders label with labelAttributes applied', () => {
    const label = 'A label'

    mount(KInput, {
      props: {
        label,
        labelAttributes: {
          info: 'some info text',
        },
      },
    })

    cy.get('.k-label').should('contain.text', label)
    cy.get('.k-label .tooltip-trigger-icon').should('exist').and('be.visible')
  })

  it('renders label and tooltip with `label-tooltip` slot applied', () => {
    const label = 'A label'

    mount(KInput, {
      props: {
        label,
      },
      slots: {
        'label-tooltip': () => h('div', {}, 'This is a tooltip'),
      },
    })

    cy.get('.k-label').should('contain.text', label)
    cy.get('.k-label .tooltip-trigger-icon').should('exist').and('be.visible')
  })

  it.skip('renders label with required symbol when `required` attribute is set', () => {
    // TODO: implement when KLabel component is reskinned
  })

  it('renders help when value is passed', () => {
    const helpText = 'I am helpful'

    mount(KInput, {
      props: {
        help: helpText,
      },
    })

    cy.get('.k-input-wrapper .help-text').should('contain.text', helpText)
  })

  it('shows character count when characterLimit prop is set and exceeded', () => {
    const textCharCount = 28
    const charLimit = 5

    mount(KInput, {
      props: {
        characterLimit: charLimit,
      },
    })

    cy.get('.k-input-wrapper .over-char-limit').should('not.exist')
    cy.get('.k-input-wrapper input.k-input').type(`This input has ${textCharCount} characters`)
    cy.get('.k-input-wrapper.has-error .help-text').should('contain.text', `${textCharCount} / ${charLimit}`)
  })

  it('reacts to text changes', () => {
    const inputValue = 'hey'
    const newValue = 'hey, dude'

    mount(KInput, {
      props: {
        modelValue: inputValue,
      },
    })

    cy.get('.k-input').should('have.value', inputValue)
    cy.get('.k-input').clear()
    cy.get('.k-input').type(newValue).then(() => {
      // Check for emitted event
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'input')
      cy.get('.k-input').should('have.value', newValue)
    })
  })

  it('renders before slot when passed', () => {
    const beforeSlot = 'before-slot'

    mount(KInput, {
      slots: {
        before: `<span data-testid="${beforeSlot}">Before slot</span>`,
      },
    })

    cy.get('.k-input-wrapper').find(`[data-testid="${beforeSlot}"]`).should('be.visible')
  })

  it('renders after slot when passed', () => {
    const afterSlot = 'after-slot'

    mount(KInput, {
      slots: {
        after: `<span data-testid="${afterSlot}">After slot</span>`,
      },
    })

    cy.get('.k-input-wrapper').find(`[data-testid="${afterSlot}"]`).should('be.visible')
  })
})
