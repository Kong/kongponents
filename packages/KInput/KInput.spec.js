import { mount } from '@vue/test-utils'
import KInput from '@/KInput/KInput'

/**
 * ALL TESTS MUST USE testMode: true
 * We generate unique IDs for reference by aria properties. Test mode strips these out
 * allowing for successful snapshot verification.
 * propsData: {
 *   testMode: true
 * }
 */
describe('KInput', () => {
  it('renders text when value is passed', () => {
    const wrapper = mount(KInput, {
      propsData: {
        testMode: true,
        value: 'Hello' // e.g. v-model
      }
    })

    expect(wrapper.find('input').element.value).toBe('Hello')
  })

  it('renders label when value is passed', () => {
    const wrapper = mount(KInput, {
      propsData: {
        testMode: true,
        label: 'A Label!'
      }
    })

    expect(wrapper.find('.k-input-label').element.innerHTML).toContain('A Label!')
  })

  it('renders label with labelAttributes applied', () => {
    const labelText = 'A Label'
    const wrapper = mount(KInput, {
      propsData: {
        testMode: true,
        label: labelText,
        labelAttributes: {
          help: 'some help text'
        }
      }
    })

    expect(wrapper.find('.k-input-label').element.innerHTML).toContain(labelText)
    expect(wrapper.find('.k-input-label .kong-icon-help').exists()).toBeTruthy()
  })

  it('renders overlayed label when value is passed', () => {
    const wrapper = mount(KInput, {
      propsData: {
        testMode: true,
        label: 'A Label!',
        overlayLabel: true
      }
    })

    expect(wrapper.find('.text-on-input label').element.innerHTML).toContain('A Label!')
  })

  it('renders small when size is passed in', () => {
    const wrapper = mount(KInput, {
      propsData: {
        testMode: true,
        size: 'small'
      }
    })

    expect(wrapper.find('.k-input-wrapper .k-input-small').exists()).toBeTruthy()
  })

  it('renders large when size is passed in', () => {
    const wrapper = mount(KInput, {
      propsData: {
        testMode: true,
        size: 'large'
      }
    })

    expect(wrapper.find('.k-input-wrapper .k-input-large').exists()).toBeTruthy()
  })

  it('renders help when value is passed', () => {
    const wrapper = mount(KInput, {
      propsData: {
        testMode: true,
        help: 'I am helpful'
      }
    })

    expect(wrapper.find('.k-input-wrapper .help').element.innerHTML).toContain('I am helpful')
  })

  it('shows a character count if the characterLimit prop is set and exceeded', () => {
    const wrapper = mount(KInput, {
      propsData: {
        testMode: true,
        characterLimit: 5
      }
    })

    expect(wrapper.find('.k-input-wrapper .over-char-limit').exists()).toBeFalsy()

    wrapper.find('input.k-input').setValue('This input has too many characters')

    expect(wrapper.find('.k-input-wrapper .over-char-limit').exists()).toBeTruthy()
  })

  it('reacts to text changes', () => {
    const wrapper = mount(KInput, {
      propsData: {
        testMode: true,
        value: 'hey'
      }
    })
    const input = wrapper.find('input')

    expect(input.element.value).toBe('hey')
    input.setValue('hey, dude')

    expect(wrapper.emitted().input).toHaveLength(1)
    expect(wrapper.emitted().input[0]).toEqual(['hey, dude'])
    expect(input.element.value).toBe('hey, dude')
  })

  it('matches snapshot', () => {
    const wrapper = mount(KInput, {
      propsData: {
        testMode: true,
        value: 'Full Name',
        placeholder: 'I am a placeholder',
        name: 'custom-input-name',
        id: 'custom-input-id',
        'data-testid': 'custom-input'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
