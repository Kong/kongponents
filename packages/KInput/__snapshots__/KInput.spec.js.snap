// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KInput matches snapshot 1`] = `
<div placeholder="I am a placeholder" name="custom-input-name" id="custom-input-id" data-testid="custom-input" class="k-input-wrapper"><input placeholder="I am a placeholder" name="custom-input-name" id="custom-input-id" data-testid="custom-input" class="form-control k-input">
  <!---->
</div>
`;
